drivers/mbox/mbox_ti_omap.c

```diff
@@ -0,0 +1,217 @@
+				.size = MAILBOX_MBOX_SIZE,
+			};
+
+			if (data->cb[i_channel]) {
+				data->cb[i_channel](dev, i_channel,
+						    data->user_data[i_channel],
+						    &msg);
+			}
+		}
+	}
+	regs->irq_regs[cfg->user_id].status_clear = flags;
+	regs->irq_regs[cfg->user_id].enable_set = irq_enabled;
+}
+
+static int omap_mailbox_send(const struct device *dev, uint32_t channel, const struct mbox_msg *msg)
+{
+	uint32_t __aligned(4) data32;
+	volatile struct omap_mailbox_regs *regs = DEV_REG_BASE(dev);
+
+	if (channel >= MAILBOX_MAX_CHANNELS) {
+		return -EINVAL;
+	}
+
+	if (msg == NULL) {
+		if (!regs->fifo_status[channel]) {
+			regs->message[channel] = 0;
+		}
+		return 0;
+	}
+
+	if (msg->size != MAILBOX_MBOX_SIZE) {
+		return -EMSGSIZE;
+	}
+
+	memcpy(&data32, msg->data, msg->size);
+
+	if (!regs->fifo_status[channel]) {
+		regs->message[channel] = data32;
+	}
+
+	return 0;
+}
+
+static int omap_mailbox_register_callback(const struct device *dev, uint32_t channel,
+					 mbox_callback_t cb, void *user_data)
+{
+	struct omap_mailbox_data *data = DEV_DATA(dev);
+
+	if (channel >= MAILBOX_MAX_CHANNELS) {
+		return -EINVAL;
+	}
+
+	data->cb[channel] = cb;
+	data->user_data[channel] = user_data;
+
+	return 0;
+}
+
+static int omap_mailbox_mtu_get(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	return MAILBOX_MBOX_SIZE;
+}
+
+static uint32_t omap_mailbox_max_channels_get(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	return MAILBOX_MAX_CHANNELS;
+}
+
+static int omap_mailbox_set_enabled(const struct device *dev, uint32_t channel, bool enable)
+{
+	struct omap_mailbox_data *data = DEV_DATA(dev);
+	const struct omap_mailbox_config *cfg = DEV_CFG(dev);
+	volatile struct omap_mailbox_regs *regs;
+	uint32_t irqstatus;
+
+	if (channel >= MAILBOX_MAX_CHANNELS) {
+		return -EINVAL;
+	}
+
+	LOG_DBG("Setting mailbox %s channel: %d", enable ? "enable" : "disable", channel);
+
+	regs = DEV_REG_BASE(dev);
+	irqstatus = regs->irq_regs[cfg->user_id].enable_set;
+	if (enable)
+		irqstatus |= MAILBOX_IRQ_NEWMSG(channel);
+	else
+		irqstatus &= ~MAILBOX_IRQ_NEWMSG(channel);
+	regs->irq_regs[cfg->user_id].enable_set = irqstatus;
+	data->channel_enable[channel] = enable;
+
+	if (enable)
+		irq_enable(cfg->irq);
+	else
+		irq_disable(cfg->irq);
+
+	return 0;
+}
+
+static const struct mbox_driver_api omap_mailbox_driver_api = {
+	.send = omap_mailbox_send,
+	.register_callback = omap_mailbox_register_callback,
+	.mtu_get = omap_mailbox_mtu_get,
+	.max_channels_get = omap_mailbox_max_channels_get,
+	.set_enabled = omap_mailbox_set_enabled,
+};
+
+#define MAILBOX_INSTANCE_DEFINE(idx)						\
+	static struct omap_mailbox_data omap_mailbox_##idx##_data;		\
+	const static struct omap_mailbox_config omap_mailbox_##idx##_config = {	\
+		DEVICE_MMIO_NAMED_ROM_INIT(reg_base, DT_DRV_INST(idx)),		\
+		.irq = DT_INST_IRQN(idx),					\
+		.user_id = DT_INST_PROP(idx, user_id),				\
+	};									\
+	static int omap_mailbox_##idx##_init(const struct device *dev)		\
```

Comment:
I still think this function should be generic. Could save some code size. Make `idx` an argument and call it here multiple times.